Introduce CMake (and removing all other project-related code)

TrueBrain opened the pull request:
Fixes #6884
Fixes #7470

[Stable release build](https://dev.azure.com/truebrain/OpenTTD/_build/results?buildId=293)
[Nightly release build](https://dev.azure.com/truebrain/OpenTTD/_build/results?buildId=294)

This removes:
- MSVC project files
- Kdevelop generation files
- configure / Makefile

It replaces them all with a single method: cmake. This should also remove any awk/vbs scripts we have around, meaning we will get a single method of doing stuff.

Currently tested with these 'IDEs':

- VS2017 (yes, VS supports CMake)
- VS2019 (yes, VS2019 supports CMake even better)
- CMake, WSL Ubuntu 18.04

This is pretty much a Work In Progress, but I wanted to show others how it would look etc, so they can also test already a bit. Any feedback is welcome.

To test it out:

```
mkdir build
cd build
cmake ..
make
```

Or open the CMakeLists.txt in VS2017 / VS2019.

Things still to do:

- [ ] Debian dependencies
- [ ] rpm support
- [ ] Linux install: menu entries
- [x] check if we are autodetecting everything
- [x] check all TODOs, and make sure they are fixed
- [x] validate this works on all current machines of all people that can compile OpenTTD now
- [x] validate this works on things like CLion, Visual Code Studio, Kdevelop, ..
- [x] rewrite Squirrel generation in CMake (done in a branch)
- [x] add the ability to make stable builds (NO_ASSERT, see MSVC changes for stable releases) via parameters
- [x] autodetect nforenum
- [x] autodetect grfcodec
- [x] add ICC support
- [x] add all the warning flags etc
- [x] add debug options
- [x] rewrite bundling code in CMake
- [x] rewrite baseset generation in CMake
- [x] rewrite regression in CMake
- [x] rewrite source.list in CMake
- [x] autodetect ICU
- [x] ~~autodetect TiMiditity~~ (libtimidity is no longer available on most OSes)
- [x] autodetect xdg-basedir
- [x] autodetect iconv
- [x] autodetect OSX stuff (Cocoa QuickDraw, Quartz, libraries, etc)
- [x] rewrite findversion in CMake (partially done here: https://github.com/LordAro/OpenTTD/commit/1ea407bfa16025fd29edbedc8c50fc27dd93d9eb)
Currently it is calling `findversion.sh`, which only works on non-Windows. Porting it to CMake makes more sense (one single way of doing stuff).
- [x] ~~add static builds~~ (no static builds for now; we might reconsider later)
- [x] add CMake support for all CI targets
- [x] revisit dependencies for Dedicated Server (does it need SDL, Cocoa, ..)
- [x] fix CI for Linux targets
- [x] fix CI for Windows
- [x] fix CI for OSX
- [x] fix mingw
- [x] fix dedicated doesn't need a video driver
- [x] split CMakeLists.txt inside several files to increase readability
- [x] remove all old project files (including configure, Makefile, projects/, etc)
- [x] restore projects\os_versions.manifest usage in VS builds

(I am sure I am missing stuff .. but this is something to work on at least :D)

glx22 commented on CMakeLists.txt:
```diff
@@ -106,10 +106,10 @@ message(STATUS "png: ${OPTION_PNG}")
-message(STATUS "Threads: ${OPTION_HAVE_THREAD}")
+set(OPTION_USE_SSE ${AUTODETECT_SSE} CACHE BOOL "Use SSE support")
+message(STATUS "SSE: ${OPTION_USE_SSE}")
+set(OPTION_USE_THREADS YES CACHE BOOL "Use threads")
```
These SSE and HAVE_THREAD changes require update in source.list else parsing fails

TrueBrain replied:
Oops :D Tnx!

nielsmh commented:
If CMake becomes the only (meta) build system, wouldn't it make sense to scrap the source.list file as well, and define that data via CMake syntax instead?

glx22 commented:
For MinGW directmusic and xaudio2 needs to be detected. MinGW has directmusic headers, but doesn't have xaudio2.

TrueBrain commented:
Added both @nielsmh  and @glx22 points to the list at the top.

orudge commented:
> For MinGW directmusic and xaudio2 needs to be detected. MinGW has directmusic headers, but doesn't have xaudio2.

Just to clarify, if you manually copy over an xaudio2 header file then mingw should be happy with it - I did test that when I wrote it.

TrueBrain commented on src/os/windows/ottdres.rc.in:
```diff
@@ -118,7 +118,7 @@ END
 
 #ifdef __MINGW32__
-1 24 "..\\..\\..\\projects\\dpi_aware.manifest"
+1 24 "${OpenTTD_SOURCE_DIR}/projects/dpi_aware.manifest"
```
Please use `CMAKE_SOURCE_DIR` here.

glx22 commented:
I kept the manifest in projects, but it should probably move somewhere else

glx22 commented:
For now we rely on cmake default CMAKE_CONFIGURATION_TYPES and CMAKE_XXXXX_FLAGS_\<CONFIG\>, but it should be better to manage that by ourselves

TrueBrain commented on src/os/windows/ottdres.rc.in:
```diff
@@ -118,7 +118,7 @@ END
 
 #ifdef __MINGW32__
-1 24 "${CMAKE_SOURCE_DIR}/projects/dpi_aware.manifest"
+1 24 "${CMAKE_SOURCE_DIR}/projects/openttd.manifest"
```
Seems like os/windows/ is a better location now :)

glx22 replied:
I will move it then, luckily only one place needs to be touched for that in cmake script ;)

TrueBrain commented on CMakeLists.txt:
```diff
@@ -179,9 +181,11 @@ if (SSE_FOUND)
-    set_source_files_properties(${SSE2_FILES} PROPERTIES COMPILE_FLAGS -msse2)
-    set_source_files_properties(${SSSE3_FILES} PROPERTIES COMPILE_FLAGS -mssse3)
-    set_source_files_properties(${SSE4_FILES} PROPERTIES COMPILE_FLAGS -msse4.1)
+    if (NOT MSVC)
```
so no need for the GLOBs either :D

glx22 replied:
good point

TrueBrain commented:
> For now we rely on cmake default CMAKE_CONFIGURATION_TYPES and CMAKE_XXXXX_FLAGS_<CONFIG>, but it should be better to manage that by ourselves

I disagree. The closer we stay to CMake, the easier our life will be. Let's try to do this, and only add exceptions if we really need to (and mark them as such). Otherwise we will have a config.lib in no time :D

glx22 commented:
Indeed something is wrong in the generated project file for lzo and lzma
AdditionalDependencies for Release:
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\lib\libpng16.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\lib\zlib.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\lzma.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\lzo2.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\lib\freetype.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\lib\bz2.lib

AdditionalDependencies for Debug:
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\libpng16d.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\lzma.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\lzo2.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\freetyped.lib
D:\developpement\GitHub\vcpkg\installed\x64-windows-static\debug\lib\bz2d.lib

glx22 commented:
hmm that's weird, OSX cmake step now pass, lzo is found, compilation fails because it doesn't find an lzo header

glx22 commented:
Ok there's an issue with lzo detection, I can compile with MinGW by luck, as the detected 
> -- LZO_INCLUDE_DIRS:H:/msys64/mingw64/include/lzo

 is covered by H:/msys64/mingw64/include from other libs.
 But on OSX 
> -- LZO_INCLUDE_DIRS: /usr/local/Cellar/lzo/2.10/include/lzo

is alone

glx22 commented:
Stupid me failing in the commit message.

BTW I think test task for MacOS should probably be "make -C build test".

nielsmh commented on cmake/AutodetectSSE.cmake:
```diff
@@ -0,0 +1,14 @@
+        set(CMAKE_REQUIRED_FLAGS "-msse4.1")
+    endif (CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
+
+    check_cxx_source_runs("
```
This may be more of a stretch goal, but won't "check runs" prevent cross compilation?

TrueBrain replied:
These things only check if it compiles. So that should just work as well for cross compilation. Pretty sure not everything in this branch cross-compiles currently, but this is one that should "just work" as far as I know. Any reason to think otherwise that we can already take into account?

nielsmh replied:
According to the manual, `check_cxx_source_runs` explicitly tests that the compiled executable also executes on the build platform. If it just needs to compile then `check_cxx_source_compiles` should be the right one, although I'm not sure if that also confirms it can link to a valid executable.

TrueBrain replied:
You are absolutely right, and this is a bug. It should be check_XXX_compiles. As indeed, running is only native :D And we just want to check if it compiles .. and it does link, so we are fine there :) Nice catch :D If you find any more, let me know!

glx22 commented:
regression works with MinGW without converting the exe, but it fails with many
```
6: -  SetCommandDelay: (null : 0x00000000)
6: +  SetCommandDelay: (null : 0x0000000000000000)'
```
I guess the lovely regex was needed :)

TrueBrain commented:
Yeah, but I refused to do that unreadable regex; so now I am just adding them one case at the time :D

glx22 commented:
Finished rewriting generate_window and squirrel_export in https://github.com/glx22/OpenTTD/commits/cmake_api

stormcone commented:
I would like to ask a little feature:

In the old './configure' way the compiled binary and the data files, which needed to run the game, are copied nicely to the 'bin' directory. Now they are in the build directory in the company of a lot of CMake generated files, what I do not really like. I tried to find a make target which copies the game file into a bundle directory, but I have not succeeded. I have some OpenTTD versions on my PC and I just run them from the 'bin' directory. Because of this I would like to avoid the use of the 'make install'.

So my request is: Could you please make a target which copies the compiled binary and the data files into a separate directory like in the old way?

janisozaur commented:
That's usually accomplished by saying where it should be installed, e.g. via `DESTDIR` environment variable.

TrueBrain commented:
You are not the first to ask @stormcone :) And @janisozaur is correct. A bit more detail:

No, we won't be copying files in `bin` anymore. Copying files in the source directory is bad practice, and was giving all kinds of unexpected issues. All created files will be in the build directory. And yes, that means between all those CMake files. But! There is a but :)
If you like, you can install OpenTTD to a random folder. This can be done with DESTDIR. For example:
`DESTDIR=$(pwd)/my-new-bin make install`
This will create `my-new-bin` with the same structure as you are used to. So this mimics both the old `bin` as the old `bundle`.
Or you can create packages, like a zip package. These are created with `make package`. You might want to disable `OPTION_INSTALL_FHS` if you want a 'flat' version (like zip, tarball, etc), instead of a debian file or something. This is what replaced the old `bundles` (not `bundle`).

If you are developing, you don't have to do any of that. You can just run `./openttd` from your build-folder, and it will work just fine. We plan on adding a `make run` target, to make that process a bit easier, but that is just a matter of convenience. You can also have several build folders around on your computer (they can be anywhere), with different version of OpenTTD. That works fine too, and is possibly more what you are looking for.

This is all standard CMake btw, at least, as far as I am aware. That means we will follow much more closely what many many other projects do, hopefully making it a lot easier for people to start helping out, also with the configuration part :)

Hope it helps! If not, let me know what use-case you are trying to solve exactly, as that was a bit vague to me ;) Tnx!

janisozaur commented:
I would advise to note it as `DESTDIR=… make install` instead of the way you did. `make` is a bit special in that it supports passing such options, but other build systems do not. `DESTDIR=… ninja install` works as intended, while the other way round gives a rather unhelpful "no such target" message.

Note, this assumes bash-like way of injecting environment variables into subprocess.

TrueBrain commented:
Both will work (for make). But I went how CMake manual tells us to: https://cmake.org/cmake/help/latest/envvar/DESTDIR.html. Properly good to tell them they should add `ninja` to that page too, and fix their darn order ;)
You are absolutely right btw :)

stormcone commented:
@janisozaur & @TrueBrain: Thank you for the detailed answers. :)

I think I will end up with the OPTION_INSTALL_FHS=OFF option and the 'make package', and a little shell script which extracts the archive into a directory. :) I do more like it than the directory structure of the DESTDIR's output.


Just a side notes: The compiling part of  the 'readme' should be updated to the new method. It is currently missing from the commits. And because I am not the first one asking this, your description should be mentioned in it in some way. :)


@TrueBrain 
> You can just run ./openttd from your build-folder, and it will work just fine.

I think currently it is not entirely true. If you want to use an AI/GS, it will not work properly, because the compat*.nut files still in the old 'bin' directory, not in the new 'build'.

glx22 commented on azure-pipelines-ci.yml:
```diff
@@ -14,22 +14,28 @@ jobs:
     displayName: 'Test'
+    inputs:
+      solution: build/RUN_TESTS.vcxproj
+      configuration: 'Release'
```
I don't understand how regression is still working, because openttd.exe is now in build/RelWithDebInfo, but regression runs in build/Release

TrueBrain committed:
Remove: all existing project files

This prepares for the switch to CMake, which takes over all current
exisiting forms of project files.

TrueBrain committed:
Codechange: move regression outside of bin and make it work via CMake script

The tst_stationlist savegame had to be changed to start the correct
AI. In the old setup, all regression AIs had the same name, which
made it impossible to run both regressions in parallel. With the new
setup this is possible.

Although all files are available to run the regression, it won't
really work till CMake is introduced (which will happen in a few
commits from here)

TrueBrain committed:
Codechange: rework how grf and ob[msg] are generated

For grfs, it now uses CMake scripts to do its job, and both grf
files are split into their own folder to make more clear what is
going on. Additionally, it no longer builds in-source (although the
resulting grf is copied back in the source folder).

For ob[msg] it now uses CMake scripts to generate the translation
files; the result is no longer stored in-source (but in the build
folder).

Although all files are available to create the GRFs and basesets, it
won't really work till CMake is introduced (which will happen in a
few commits from here)

TrueBrain committed:
Add: introduce CMake for project management

CMake works on all our supported platforms, like MSVC, Mingw, GCC,
Clang, and many more. It allows for a single way of doing things,
so no longer we need shell scripts and vbs scripts to work on all
our supported platforms.

Additionally, CMake allows to generate project files for like MSVC,
KDevelop, etc.

This heavily reduces the lines of code we need to support multiple
platforms from a project perspective.

Addtiionally, this heavily improves our detection of libraries, etc.

TrueBrain committed:
Add: create bundles via CPack

CPack works closely together with CMake to do the right thing in
terms of bundling (called 'package'). This generates all the
packaging we need, and some more.

TrueBrain committed:
Fix: [AzurePipelines] switch the CI / CD to CMake too

This also means dropping Debian/jessie, as it has a CMake that is
too old (3.0), with no real path to upgrade.

TrueBrain committed:
Codechange: remove #ifdef from .cpp files to exclude features

With CMake, these files are simply not compiled to start with.

TrueBrain committed:
Fix: update COMPILING.md stating what changed with CMake

TrueBrain commented:
I squashed the whole PR from 200+ commits to 8. This makes the PR a lot more reviewable. There are still a few odds and ends that needs to be connected, but most of the features are now in there.

glx22 commented on azure-pipelines/templates/release.yml:
```diff
@@ -73,32 +73,30 @@ jobs:
-        make -f Makefile.msvc bundle_exe PLATFORM=$(BundlePlatform) BUNDLE_NAME=openttd-$(Build.BuildNumber)-windows-$(BundlePlatform)
-      displayName: 'Create installer bundle'
+
+      cd build/RelWithDebInfo/openttd.pdb build/bundles/openttd-$(Build.BuildNumber)-windows-$(BundlePlatform).pdb
```
Should be `cp` not `cd`.

glx22 commented:
I think os/windows/openttd.manifest should be in "Add: introduce CMake for project management" and not in "Add: create bundles via CPack" as it's needed for compilation

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+
+if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+    message(FATAL_ERROR "In-source builds not allowed. Please run \"cmake ..\" from the bin directory")
+endif (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
```
The arguments for else() and endif() are optional, and are usually not set as it is just another place that needs to be touched if the if() is ever changed.

TrueBrain replied:
Indeed, it is optional. But we noticed (and also in other projects) that it avoids a lot of common mistakes. So I personally really like this approach, and am tempted to keep it :)

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake")
+
+include(Options REQUIRED)
```
include() has an OPTIONAL argument, but no REQUIRED. I wonder why this works at all.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+# Make sure we have Threads available. We don't use autodetect_library here,
+# as the CMake 'FindThreads' is a lot more clever, always available, and
+# gives us a Threads::Threads dependency to link to our application.
+find_package(Threads REQUIRED)
```
```cmake
set(CMAKE_THREAD_PREFER_PTHREAD True)
```
before calling this to get the proper -pthread flag.

TrueBrain replied:
I am not sure; do we prefer pthread? Especially now we use C++ `std::thread`, does it matter? (honest question). @michicc / @LordAro : do you happen to have any insight here?

DerDakon replied:
You usually need it, and you want the -pthread flag instead of just using -lphtread with the linker. There be dragons.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+find_package(Threads REQUIRED)
+
+include(AutodetectLibrary REQUIRED)
+autodetect_library("zlib" ZLIB ZLIB zlib "" zlib.h z)
```
CMake has modules that care for libz, png and a few others, that already know all the header and library names. I would not duplicate it.

TrueBrain replied:
Yes and no. It does for some libraries, it does not for others. To avoid making some library detection more special than others, we decided to just all make them equal. In reality, it will never fall back on pkg-config or manual detection for these anyway, so there is no real harm.

Mostly we noticed this with libraries that are supported in newer CMake, but not in older. There it became really blurry real quick. Given the wide range of systems we need to support, to me this feels like the most safe solution.

I am open for counter-arguments ;)

LordAro replied:
AIUI, it is more common to provide your own `Find*` functions, if one doesn't exist in cmake itself

e.g. https://github.com/SFTtech/openage/tree/master/buildsystem/modules

DerDakon replied:
I would go this route, too. This is easier to follow for everyone used to CMake, and you still can call your custom detect function in your own Find*.cmake if that really makes things easier.

LordAro replied:
I'd like to see this addressed before merge. There's no reason to go all NIH for library detection, of all things.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+
+# IPO is only properly supported from CMake 3.9. Despite the fact we are
+# CMake 3.5, still enable IPO if we detect we are 3.9+.
+if (NOT CMAKE_VERSION VERSION_LESS 3.9)
```
You can simply do: 
```cmake
if(POLICY CMP0069)
```

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+    endif (NOT SDL_FOUND)
+endif (UNIX AND NOT APPLE AND NOT OPTION_DEDICATED)
+if (APPLE)
+    if (NOT AUDIOTOOLBOX_LIBRARY)
```
just add REQUIRED to all the find_library() calls of this Apple things above.

TrueBrain replied:
`find_library()` doesn't have `REQUIRED`, according to the manual? (am I missing something here?)
https://cmake.org/cmake/help/latest/command/find_library.html

DerDakon replied:
Ah, yes, I was at find_package().

DerDakon replied:
It will happen with CMake 3.18 ;)

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+# All other files are added via target_sources()
+
+include(AddCustomXXXTimestamp REQUIRED)
+add_subdirectory(${CMAKE_SOURCE_DIR}/src)
```
the directories are always relative to the current source directory, so this can be just "src".

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+if (APPLE)
+    link_library(Iconv)
+
+    target_link_libraries(openttd ${AUDIOTOOLBOX_LIBRARY})
```
You can do all of them in a single call if you like.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+    target_link_libraries(openttd ${CARBON_LIBRARY})
+    target_link_libraries(openttd ${COCOA_LIBRARY})
+
+    add_definitions(-DWITH_COCOA)
```
You can do all of them in a single call if you like.

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+endif (WIN32)
+
+if (CMAKE_SIZEOF_VOID_P EQUAL 8)
+    add_definitions(-D_SQ64)
```
I have not looked into the actual code, but can't you just check this there? Even if it is an ordinary
```c
if(sizeof(void*) == 8)
```
the compiler will optimize it away as it will detect that it will not change at runtime.

TrueBrain replied:
Yes, we could; but Squirrel is a 3rdparty library, so these kind of changes can be hefty. It is simply easier to do it like this. But for sure this needs further attention towards the future :)

DerDakon commented on cmake/AddCustomXXXTimestamp.cmake:
```diff
@@ -0,0 +1,145 @@
+        # Check if the output is produced by add_custom_command_timestamp()
+        get_source_file_property(BYPRODUCT ${DEPEND} BYPRODUCT)
+
+        if ("${BYPRODUCT}" STREQUAL "NOTFOUND")
```
```cmake
if(BYPRODUCTS STREQUAL "NOTFOUND")
```

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,237 @@
+
+project(OpenTTD)
+
+if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
```
```cmake
if(CMAKE_SOURCE_DIR STREQUAL CMAKE_BINARY_DIR)
```

CMake will then expand the variables automatically before checking.

DerDakon commented on cmake/AutodetectEditbin.cmake:
```diff
@@ -0,0 +1,29 @@
+    message(STATUS "Detecting editbin")
+    find_program(EDITBIN_EXECUTABLE editbin.exe)
+
+    if (EDITBIN_EXECUTABLE STREQUAL "EDITBIN_EXECUTABLE-NOTFOUND")
```
```
if (NOT EDITBIN_EXECUTABLE)
```
But I would always calculate the MSVC path and just pass it to find_program() in HINTS.

DerDakon commented on cmake/AutodetectLibrary.cmake:
```diff
@@ -0,0 +1,164 @@
+        pkg_search_module("${NAME}" "${PKGCONFIG}" QUIET)
+
+        # pkg-config returns the libraries without really resolving them.
+        # This has as huge drawback that sometimes you only notice on linking
```
Maybe you should require a more recent CMake version: newer versions of FindPkgConfig do exactly that. And they return all things like library, compile flags, and include directories in an imported target.

TrueBrain replied:
I wish we could. I looked for the highest CMake version we needed to support (for now), which is 3.5. But over time we can bump it, and remove these quirks for sure :) Happy to see it is fixed in more recent versions, as it was really weird :P

DerDakon replied:
What's the problem? Kitware provides statically linked binaries for at least Windows and Linux and an Ubuntu repo. Someone wants to build your thing with a ton of external dependencies, getting a recent CMake version shoudn't be a big deal.

DerDakon commented on cmake/AutodetectLibrary.cmake:
```diff
@@ -0,0 +1,164 @@
+endmacro()
+
+macro(_patch_older_cmake NAME)
+    # Patch up libraries that still announce under the deprecated name only
```
Just require a newer version of CMake. If you still find modules that behave like this file a bug.

DerDakon commented on cmake/AutodetectLibrary.cmake:
```diff
@@ -0,0 +1,164 @@
+                    string(REPLACE "/lib/" "/debug/lib/" PATCH_LIBRARY_DEBUG "${PATCH_LIBRARY}")
+                endif (${PATCH_LIBRARY} MATCHES "/debug/")
+
+                LIST(APPEND PATCH_LIBRARIES
```
cmake --help-module SelectLibraryConfigurations

TrueBrain replied:
The issue here really seems to be buggy stuff in `vcpkg`. Sometimes we see libraries that only return `debug` libraries, and other times only `optimized`. Not sure how `select_library_configurations()` is going to help here? (I am happy to be shown wrong btw; this is really annoying code to have, and to maintain).

DerDakon replied:
it can help setting the complete library variable from *_RELEASE and *_DEBUG variables.

glx22 replied:
And the issue is only when generating a MSVC project (multi configuration), single configuration using Ninja generator works as intended.

DerDakon replied:
The problem exists only for MSVC as it has distinct C++ runtimes for release and debug.

glx22 replied:
Ninja builds using MSVC compiler too, but vcpkg returns the correct debug/release lib in this case as only one kind is searched. It's really only when both kind are wanted that vcpkg fails, and not for all libs.

DerDakon commented on cmake/SourceList.cmake:
```diff
@@ -0,0 +1,63 @@
+        list(GET ENTRY 0 FILE)
+        list(GET ENTRY 1 PROPERTIES)
+
+        set_source_files_properties(${FILE} PROPERTIES COMPILE_FLAGS ${PROPERTIES})
```
Have you tried set_property(SOURCE …)?

TrueBrain replied:
Aren't they identical? Does it matter? (honest question again)

DerDakon replied:
It could help by setting these this from different locations. Maybe. I was told it is more flexible, e.g. it has APPEND. Just test if it makes your use case easier.

TrueBrain commented:
Wow, a lot to unpack here :) Tnx a lot @DerDakon! Sadly, GitHub only allows me to reply on one thing at the time, so this might get a bit spammy ;) But I seriously appreciate the review!

DerDakon commented on cmake/PackageBundle.cmake:
```diff
@@ -0,0 +1,24 @@
+# Create a temporary Info.plist.in, where we will fill in the version via
+# CPackProperties.cmake.in. This because at this point in time the version
+# is not yet known.
+configure_file("${CMAKE_SOURCE_DIR}/os/macosx/Info.plist.in" "${CMAKE_CURRENT_BINARY_DIR}/Info.plist.in")
```
Maybe you want @ONLY.

TrueBrain commented:
In my opinion, this is now ready to be merged. At the top are now two URLs that point to successful release builds; so nightlies etc will also just work after merging.

There are a few things left, on which I need help, all to do with packaging:
 - Linux 'install' can prepare menu entries, but I haven't figured out how yet.
 - Debian dependencies are wrong, and needs tuning.
 - RPM cannot be produced currently.
For this I suggest to make a new ticket to track this, as they are only really important to fix before 1.10 is released.

Otherwise it is completely done, from what I can tell.

PeterN commented:
Should the AI/GS compat files be copied into the appropriate directory automatically as well?

glx22 commented:
I think the current way to run is to do `../build/openttd` from `<repo>/bin` directory. We set `VS_DEBUGGER_WORKING_DIRECTORY` property to `<repo>/bin` so MSVC debugger does that automatically.

BTW I noticed, when installing openttd from MSVC, that InstallAndPackage.cmake is not selective enough for bin/ai and bin/game. It copies all files/dir, even if they are not part of the clone.

glx22 commented:
When doing the rebase I decided to conditionally exclude some library detection for windows as FreeType takes precedence over GDI, and its development files can be present on the system even if we don't want to use it.

fsimonis commented:
I may be a _bit late_ to the party, but I started a CMake port of OpenTTD about one year ago. After reading his Trello board, I attempted to contact @TrueBrain back then via Gitter to join forces, but that sadly never happened. You can find my version on [my fork](https://github.com/fsimonis/OpenTTD/tree/cmake-port).  
It runs on VS and various Linux distros. I couldn't test on Mac due to a lack of hardware.

I stopped the port when I hit the wall of supporting all the OSes. Some of which CMake isn't even aware of.

Some lessons I have learned and general feedback to this port:

1. autodetection of libraries is a total nightmare for package maintainers as one never really knows what the actual configuration looks like. Also, the maintainer won't see errors when the configuration did not find all expected dependencies.
The clean way to handle this in CMake is to use options for every dependency (e.g. `OPENTTD_ALLEGRO`), always require the enabled ones and enable only the most general/cross-platform libraries by default.  
One can then use the [CMake option to pre-populate the cache](https://cmake.org/cmake/help/latest/manual/cmake.1.html#options) to load a configuration for a specific OS `cmake -C linux-ubuntu-1804.cmake`. This makes it also easier to keep the dependencies in sync with package dependencies such as Debian. Most importantly: this solution is super clean!

2. As OpenTTD is a distributed executable, there is no reason to stick to an old CMake version. Installing CMake is as trivial as downloading, extracting an archive and maybe extending `PATH`. The end-user will never need to install CMake anyhow.
Staying up2date has 3 major upsides:

    2. The FindModules are up2date and _mostly_ provide targets. (You can find the missing modules on my fork)
    3. The general support for targets is better.
    4. The support for installing various file types is far better via the `TYPE` argument of `install`.

3. Using targets to link dependencies allows to write sane CMake scripts. Everything else just silently fails and is a nightmare to debug.

4. The third-party libraries `squirrel`, `squirrel-stl`, `os2`, and `md5` can be separate targets. (The common `string` dependency is a nuisance, but not really a problem.)

5. Prefixing all project-related options and cache variables with `OPENTTD_` makes them easier to find/grep, will neatly align them in `ccmake` and group them in `cmake-gui`.

__Thank you for doing this port!__
The build system and SVN were the two reasons that always kept me from contributing to this project!

DerDakon commented:
Regarding your first concern, I suggest having a look at ```cmake --help-module FeatureSummary```. I would very much suggest _not_ to maintain a list of preconfigured caches. Either CMake is able to detect everything from distro packages, or you file a bug against CMake. Everything else just papers over existing bugs.

I agree on 2 and 3, and have no opinion about the others.

DerDakon commented on .gitignore:
```diff
@@ -21,18 +21,13 @@ bin/scripts/*
 docs/source/*
-.kdev4
-.kdev4/*
-*.kdev4
```
I would keep at least *.kdev4, this would still be relevant when importing the CMake project into KDevelop.

glx22 replied:
That can be handle in the local gitignore. I see no real need to put it globally.

DerDakon commented on CPackProperties.cmake.in:
```diff
@@ -0,0 +1,12 @@
+# Make the current version available to CPack
+set(CPACK_PACKAGE_VERSION "@REV_VERSION@")
```
It should be enough if you call ```project(OpenTTD VERSION 5.6.7)``` at the top level, that would make the variable ```OpenTTD_VERSION``` available, also to CPack.

fsimonis replied:
The `REV_VERSION` is based on the git tag(project version) or the git commit and can change between builds.
This is not pretty but necessary to make CPack behave correctly.

DerDakon replied:
Please not that this will usually not change between builds: CMake will not notice if your git status has changed, so it will not rerun CMake, so it will not rerun your ```execute_process()```.

DerDakon commented on cmake/AutodetectEditbin.cmake:
```diff
@@ -0,0 +1,26 @@
+# autodetect_editbin()
+#
+function(autodetect_editbin)
+    if (DEFINED EDITBIN_FOUND)
```
You could just check on EDITBIN_EXECUTABLE here, CMake will automatically put that into cache in find_program(). And apart from the messages you do not need to guard this, if the variable is already set find_program() will just do nothing.

fsimonis replied:
We could refactor this entire function into very small and simple find module. Then one could use the more CMakeish `find_package(EDITBIN)`.
You can have a look at the [`FindCCACHE`](https://github.com/fsimonis/OpenTTD/blob/cmake-port/cmake/FindCCACHE.cmake) on my fork for a minimal example.
This also keeps the error messages consistent with other find modules.

LordAro replied:
This should be addressed too

DerDakon commented on cmake/AutodetectEditbin.cmake:
```diff
@@ -0,0 +1,26 @@
+    message(STATUS "Detecting editbin - found")
+
+    # Make sure these values are cached properly
+    set(EDITBIN_FOUND YES CACHE INTERNAL "")
```
If you want to have this variable after all simply not cache it: the find_program() call will do nothing, EDITBIN_EXECUTABLE is set, so this variable get's set to YES anyway. No need to clutter the cache with it.

DerDakon commented on cmake/AutodetectLibrary.cmake:
```diff
@@ -0,0 +1,164 @@
+
+    # Check if any of the argument is ENCOURAGED
+    list(FIND args ENCOURAGED ENCOURAGED)
+    if (ENCOURAGED LESS 0)
```
```cmake
if("ENCOURAGED" IN_LIST args)
```

DerDakon commented on cmake/AutodetectSSE.cmake:
```diff
@@ -0,0 +1,20 @@
+    include(CheckCXXSourceCompiles)
+    set(CMAKE_REQUIRED_FLAGS "")
+
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
```
This is missing a check if this is an x86 platform at all. I would just go and look if the flag is supported or not, and just build without it if not.

DerDakon commented on cmake/CompileFlags.cmake:
```diff
@@ -0,0 +1,120 @@
+            -wd2160
+        )
+    else ()
+        message(FATAL_ERROR "No warning flags are set for this compiler yet; please consider creating a Pull Request to add support for this compiler.")
```
This should be downgraded to a warning, if the compiler is unknown just go on and look what happens.

DerDakon commented on cmake/CreateGrfCommand.cmake:
```diff
@@ -0,0 +1,50 @@
+        get_filename_component(GRF_PNG_SOURCE_FILE_NAME "${GRF_PNG_SOURCE_FILE}" NAME)
+        set(GRF_PNG_BINARY_FILE "${CMAKE_CURRENT_BINARY_DIR}/sprites/${GRF_PNG_SOURCE_FILE_NAME}")
+
+        add_custom_command(OUTPUT ${GRF_PNG_BINARY_FILE}
```
you could simply use ```configure_file(... COPYONLY)```, at the expense that this will happen at configure time instead of build time, but otherwise works the same.

DerDakon commented on cmake/CreateGrfCommand.cmake:
```diff
@@ -0,0 +1,50 @@
+
+    add_custom_command(OUTPUT ${GRF_BINARY_FILE}
+            COMMAND ${CMAKE_COMMAND}
+                    -DGRF_SOURCE_FOLDER=${CMAKE_CURRENT_SOURCE_DIR}
```
I wonder if that thing can't simply be told to read the PNGs from the source directory, that would avoid the entire copying.

DerDakon commented on cmake/InstallAndPackage.cmake:
```diff
@@ -0,0 +1,109 @@
+# If requested, use FHS layout; otherwise fall back to a flat layout.
+if (OPTION_INSTALL_FHS)
+    set(BINARY_DESTINATION_DIR "bin")
```
Please just scrap that and use the GNUInstallDirs module. Packagers will love you for that.

DerDakon commented on cmake/Options.cmake:
```diff
@@ -0,0 +1,84 @@
+        set(DEFAULT_SHARED_DIR "(not set)")
+        set(DEFAULT_GLOBAL_DIR "${CMAKE_INSTALL_PREFIX}/share/games/openttd")
+    else ()
+        message(FATAL_ERROR "Unknown OS found; please consider creating a Pull Request to add support for this OS.")
```
Again, just leave it as is. Let the *BSD folks do their experiment…

DerDakon commented on cmake/SourceList.cmake:
```diff
@@ -0,0 +1,63 @@
+        endif (NOT (${PARAM_CONDITION}))
+    endif (PARAM_CONDITION)
+
+    foreach(FILE ${PARAM_FILES})
```
IN LISTS

DerDakon commented on cmake/SourceList.cmake:
```diff
@@ -0,0 +1,63 @@
+    set(PARAM_FILES "${PARAM_UNPARSED_ARGUMENTS}")
+
+    if (PARAM_CONDITION)
+        if (NOT (${PARAM_CONDITION}))
```
Why this? I don't think you can ever reach the inner ```return()``` as an empty variable is just false.

DerDakon commented on cmake/SourceList.cmake:
```diff
@@ -0,0 +1,63 @@
+
+    get_property(SOURCE_PROPERTIES GLOBAL PROPERTY source_properties)
+
+    foreach(FILE ${PARAM_FILES})
```
IN LISTS

DerDakon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,254 @@
+
+if (CMAKE_SOURCE_DIR STREQUAL CMAKE_BINARY_DIR)
+    message(FATAL_ERROR "In-source builds not allowed. Please run \"cmake ..\" from the bin directory")
+endif (CMAKE_SOURCE_DIR STREQUAL CMAKE_BINARY_DIR)
```
I probably have said this before, but just remove the arguments to all end*() calls. It makes code much easier to maintain.

fsimonis commented:
@DerDakon I have the feeling that we are talking past each other.
I am talking about using the precaches solely for explicitly en/disabling features. This matter is orthogonal to how those features are detected. I totally agree that using the precache to influence library detection is a very _very_ bad idea.

### Explicit dependencies vs autodetection

My actual concern is that we should use explicit bindings options for dependencies. Autodetecting every library makes life more difficult for packagers and developer.

Here my reasoning:
1. With hard autodetection, CMake will _always_ configure the version with the most detected libraries.
2. CMake will configure successfully even if an installed library was not found. A developer thus always has to check the logs to validate that all available libraries were detected as expected.
3. As a developer you may want to debug a configuration with a specific feature-set to reproduce or locate a bug. Hard autodetection makes this impossible.
4. In CI you may want to run tests on OpenTTD with a fixed feature set, such as only a single video and audio driver enabled or even multiple configuration matrices.

I propose to do the following:
1. Add a boolean option for every external library.
2. Explain why we need it. (@DerDakon Thanks, I was not aware of this module!)
3. Always fail if the library was not found.

Example for zlib:
```cmake
# Options
option(OPENTTD_ZLIB "Enable ZLib" ON`)
add_feature_info(ZLib OPENTTD_ZLIB "Required for (de)compressing of old (0.3.0-1.0.5) savegames, content downloads, and heightmaps.")

# ...

# Package discovery
if(OPENTTD_ZLIB)
  find_package(ZLIB REQUIRED)
endif()

# ...
```

Running CMake without a preset of features will likely fail during the first configuration attempt, but then provide a full list of options to `ccmake`/`cmake-gui` etc.

### Feature set using pre-cache population
Example: lets consider the different compression libraries used in OpenTTD. Then the options and the defaults for those features could look as follows:

| Library | CMake Option | Default |
| --- | --- | --- |
| zlib | `OPENTTD_ZLIB` | `ON` |
| LIBLZO2 | `OPENTTD_LIBLZO2` | `ON` |
| libzlma | `OPENTTD_LIBZLMA` | `ON` |

So for Ubuntu 18.04 we know that all libraries are available, thus we could explicitly enable them as such:
```cmake
# linux-ubuntu-1804.cmake
set(OPENTTD_ZLIB ON CACHE BOOLEAN "")
set(OPENTTD_LIBLZO2 ON CACHE BOOLEAN "")
set(OPENTTD_LIBZLMA ON CACHE BOOLEAN "")
```

On some other system, `liblzio2` may not be available, thus we could explicitly disable it:
```cmake
# some-other-system.cmake
set(OPENTTD_ZLIB ON CACHE BOOLEAN "")
set(OPENTTD_LIBLZO2 OFF CACHE BOOLEAN "")
set(OPENTTD_LIBZLMA ON CACHE BOOLEAN "")
```

This way, the CI configuration for a specific distro boils down to: `cmake -C $SRC/.../linux-ubuntu-1804.cmake $SRC`.

DerDakon commented:
I would use options for feature sets, and based on that require on or another library. Even if Ubuntu has Lzo2, it doesn't mean that the user has it installed. What you could do is to give the user a list of packages for a given distro that should ideally be installed.

fsimonis commented:
So a possible feature-set based on the [README](https://github.com/OpenTTD/OpenTTD#71-requiredoptional-libraries) would look like this:

Feature option | Required libraries
| --- | --- |
`OPENTTD_BACKEND_SDL`      | libsdl
`OPENTTD_BACKEND_ALLEGRO`  | liballegro
`OPENTTD_SAVEGAME_ANCIENT` | liblzo2
`OPENTTD_SAVEGAME_OLD`     | zlib
`OPENTTD_SAVEGAME`         | liblzma
`OPENTTD_CONTENT_DOWNLOAD` | zlib
`OPENTTD_HEIGHTMAP`        | zlib, libpng
`OPENTTD_SCREENSHOT`       | libpng
`OPENTTD_FONTS`            | libfreetype, libfontconfig
`OPENTTD_RIGHT_TO_LEFT`    | libico

@DerDakon Is that what you had in mind?

DerDakon commented:
Basically. And the code for that could look like this:

```cmake
unset(zlib_required)
option(OPENTTD_SAVEGAME_OLD "foo" Off)
option(OPENTTD_HEIGHTMAP "bar" On)

if (OPENTTD_HEIGHTMAP OR OPENTTD_SAVEGAME_OLD OR …)
  set(zlib_required "REQUIRED")
endif()

find_package(ZLIB ${zlib_required})
```

glx22 commented:
I think I properly handled integration of 2d27e8e68 and c655f89d8

nielsmh commented on cmake/InstallAndPackage.cmake:
```diff
@@ -0,0 +1,109 @@
+        COMPONENT language_files)
+
+install(FILES
+                ${CMAKE_SOURCE_DIR}/COPYING
```
```suggestion
                ${CMAKE_SOURCE_DIR}/COPYING.md
```

nielsmh commented on cmake/InstallAndPackage.cmake:
```diff
@@ -0,0 +1,109 @@
+                ${CMAKE_SOURCE_DIR}/COPYING
+                ${CMAKE_SOURCE_DIR}/README.md
+                ${CMAKE_SOURCE_DIR}/changelog.txt
+                ${CMAKE_SOURCE_DIR}/docs/multiplayer.txt
```
```suggestion
                ${CMAKE_SOURCE_DIR}/docs/multiplayer.md
```

nielsmh reviewed (commented):
Two files have been renamed

nielsmh commented:
Unexpected but nice bonus: VS2019 also allows installing clang-cl easily, and it Just Works with CMake. The final EXE file is smaller _and_ faster than the one build with Microsoft's compiler. I got almost 10% faster frame times on the Wentbourne save.

LordAro commented on CMakeLists.txt:
```diff
@@ -0,0 +1,255 @@
+endif (NOT GLOBAL_DIR STREQUAL "(not set)")
+
+if (SSE_FOUND)
+    add_definitions(-DWITH_SSE)
```
Should probably be consistent here and split this like the others

LordAro commented on azure-pipelines/templates/release.yml:
```diff
@@ -73,33 +73,33 @@ jobs:
       Win64:
-        BuildPlatform: 'x64'
         BundlePlatform: 'win64'
+        OptionalBuildArch: ' Win64'
```
leading space here is weird?

LordAro commented on cmake/AddCustomXXXTimestamp.cmake:
```diff
@@ -0,0 +1,145 @@
+    endforeach(MULTI)
+endmacro()
+
+# Generated files can be older than their dependencies, causing useless
```
Why is this necessary? Under what circumstances can generated files be older than their dependencies? That doesn't make sense at all.

glx22 replied:
When a dependency is modified (like the cmake script used to generate a file), the change may have no effect to the generated file, because `configure_file()` basically creates a temp file and updates the generated file only if the temp file is different.
Same happens with `strgen` and `settingsgen`, they update the table headers only if they have changed.
So yes, generated files can be older than their dependencies :)

LordAro reviewed (dismissed):
Basically, all the comments by @DerDakon need addressing, IMO :)

DerDakon commented on cmake/CompileFlags.cmake:
```diff
@@ -0,0 +1,120 @@
+        )
+
+        if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+            include(AutodetectGCCCompileFlag)
```
What about simply using CheckCCompilerFlag or CheckCXXCompilerFlag?

DerDakon commented on CMakeLists.txt:
```diff
@@ -12,6 +12,9 @@ include(Options)
 set_directory_options()
 
+include(Static)
+set_static()
```
I don't like the name as it is misleading for every platform but MinGW. Either use a better name or also guard it with an ```if (MINGW)``` here so it becomes obvious that it does nothing otherwise.

DerDakon commented on cmake/FindFontconfig.cmake:
```diff
@@ -0,0 +1,75 @@
+FindFontconfig
+-------
+
+Finds the fontconfig library.
```
Scrap that thing ;) Either just require CMake 3.14 (which has FindFontconfig), or copy that module from current CMake.

DerDakon commented on cmake/FindFluidsynth.cmake:
```diff
@@ -0,0 +1,65 @@
+find_package(PkgConfig QUIET)
+pkg_check_modules(PC_Fluidsynth QUIET fluidsynth)
+
+find_path(Fluidsynth_INCLUDE_DIR
```
Either specify IMPORTED_TARGET and just use the the target or use the PC_Fluidsynth_LINK_LIBRARIES variable for the libs (since 3.12). The PC_Fluidsynth_INCLUDE_DIRS variable cn be used directly.

DerDakon commented on cmake/LinkLibrary.cmake:
```diff
@@ -0,0 +1,22 @@
+    # Check if any of the argument is ENCOURAGED
+    if ("ENCOURAGED" IN_LIST args)
+        set(ENCOURAGED YES)
+    else ("ENCOURAGED" IN_LIST args)
```
drop the arguments from ```else()``` and ```endif()```, they only add clutter and maintenance burden.

DerDakon commented on cmake/LinkLibrary.cmake:
```diff
@@ -0,0 +1,22 @@
+function(link_library NAME)
```
This is missing function documentation.

DerDakon commented on cmake/LinkLibrary.cmake:
```diff
@@ -0,0 +1,22 @@
+        string(TOUPPER "${NAME}" UCNAME)
+        add_definitions(-DWITH_${UCNAME})
+        include_directories(${${NAME}_INCLUDE_DIRS})
+        target_link_libraries(openttd ${${NAME}_LIBRARIES})
```
I would allow to pass a target to this function. If a target is given then just link that, and only otherwise use ```include_directories()``` and link to the *_LIBRARIES variable. Targets can transport additional things like link flags, compile definitions and so on and should generally be preferred.

fsimonis replied:
@DerDakon I agree and addressed this in https://github.com/TrueBrain/OpenTTD/pull/11

fsimonis commented:
How would you like to specify the Debian/RPM dependencies to use during package generation? I can think of 2 options:

1. Detect the distribution and release using `lsb_release` and set `CPACK_DEBIAN_PACKAGE_DEPENDS` as well as the package name based on that. This way, the user/CI could simply call `make package` in e.g. an `ubuntu:18.04` docker container and the resulting package would be ready to use.
2. Let the user/CI pass the `CPACK_DEBIAN_PACKAGE_DEPENDS` and maybe the package name to cmake. This would decouple the distribution-based dependency management from the openttd code.

nielsmh commented:
Unless someone considers the three missing checkboxes vital, I think this is ready to get merged? It seemed to work well enough last I tried things with VS2019.

James103 commented:
@nielsmh Not unless the "Temp: debug output" commit is reworded and the OpenGFX installation procedure is fixed (currently fails despite #8105 merge).

James103 commented:
Conflicts still need to be resolved (`media/baseset/translations.vbs`) and there's still the `Temp: debug output` commit that needs rewording.

glx22 commented:
The Temp commit is here for a reason, and the conflict will go with the next rebase.

fsimonis commented:
Hi, please check out https://github.com/TrueBrain/OpenTTD/pull/11 before you merge this PR.
The PR adds potential targets to the library detection and fixed an incompatibility with FindSDL.

glx22 commented:
@fsimonis I implemented your PR when you first mentioned it :)

nielsmh reviewed (dismissed):
I'm going to be radical and say we should get this merged :) Iron out the final kinks after we see it in production. Make sure we don't get any more PRs wanting to change things about the build system whose work would just have to be duplicated.

Well, of course there's a bunch of formal errors from the commit checker that need to be fixed, and some "conflicts".

glx22 commented:
Fixed conflicts, removed temp commit, and squashed

nielsmh reviewed (approved):
Let's do this!

LordAro reviewed (approved):
Screw it.
